Factory function and basic .sizes() support for C++ NestedTensor

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #117905
* #119977
* #119976
* #121481

This PR:
- updates backend select codegen to also check for keys in singleton ints
- updates C++ NestedTensor to return sizes using singleton ints
- Adds support for C++ NestedTensor factory function to create NestedTensor using metadata on singleton int. Only supports a single ragged dimension.
- Unifies codepath for a particular backward formula to test this

soulitzer committed:
Basic .sizes() support for C++ NestedTensor

[ghstack-poisoned]

soulitzer committed:
Update on "Factory function and basic .sizes() support for C++ NestedTensor"




[ghstack-poisoned]

soulitzer committed:
Update on "Factory function and basic .sizes() support for C++ NestedTensor"


This PR:
- updates backend select codegen to also check for keys in singleton ints
- updates C++ NestedTensor to return sizes using singleton ints
- Adds support for C++ NestedTensor factory function to create NestedTensor using metadata on singleton int. Only supports a single ragged dimension.
- Unifies codepath for a particular backward formula to test this

TODO:
- update tests that previously verify that returning sizes errors
- Check that we don't regress split_with_sizes if it previously supported multiple ragged dimensions (possibly just add support for this)


[ghstack-poisoned]

soulitzer committed:
Update on "Factory function and basic .sizes() support for C++ NestedTensor"


This PR:
- updates backend select codegen to also check for keys in singleton ints
- updates C++ NestedTensor to return sizes using singleton ints
- Adds support for C++ NestedTensor factory function to create NestedTensor using metadata on singleton int. Only supports a single ragged dimension.
- Unifies codepath for a particular backward formula to test this

TODO:
- update tests that previously verify that returning sizes errors
- Check that we don't regress split_with_sizes if it previously supported multiple ragged dimensions (possibly just add support for this)


[ghstack-poisoned]

soulitzer committed:
Update on "Factory function and basic .sizes() support for C++ NestedTensor"


This PR:
- updates backend select codegen to also check for keys in singleton ints
- updates C++ NestedTensor to return sizes using singleton ints
- Adds support for C++ NestedTensor factory function to create NestedTensor using metadata on singleton int. Only supports a single ragged dimension.
- Unifies codepath for a particular backward formula to test this

TODO:
- update tests that previously verify that returning sizes errors
- Check that we don't regress split_with_sizes if it previously supported multiple ragged dimensions (possibly just add support for this)


[ghstack-poisoned]

soulitzer committed:
Update on "Factory function and basic .sizes() support for C++ NestedTensor"


This PR:
- updates backend select codegen to also check for keys in singleton ints
- updates C++ NestedTensor to return sizes using singleton ints
- Adds support for C++ NestedTensor factory function to create NestedTensor using metadata on singleton int. Only supports a single ragged dimension.
- Unifies codepath for a particular backward formula to test this

TODO:
- update tests that previously verify that returning sizes errors
- Check that we don't regress split_with_sizes if it previously supported multiple ragged dimensions (possibly just add support for this)


[ghstack-poisoned]

soulitzer committed:
Update on "Factory function and basic .sizes() support for C++ NestedTensor"


This PR:
- updates backend select codegen to also check for keys in singleton ints
- updates C++ NestedTensor to return sizes using singleton ints
- Adds support for C++ NestedTensor factory function to create NestedTensor using metadata on singleton int. Only supports a single ragged dimension.
- Unifies codepath for a particular backward formula to test this

TODO:
- update tests that previously verify that returning sizes errors
- test that we support multiple ragged dimensions


[ghstack-poisoned]

soulitzer committed:
Update on "Factory function and basic .sizes() support for C++ NestedTensor"


This PR:
- updates backend select codegen to also check for keys in singleton ints
- updates C++ NestedTensor to return sizes using singleton ints
- Adds support for C++ NestedTensor factory function to create NestedTensor using metadata on singleton int. Only supports a single ragged dimension.
- Unifies codepath for a particular backward formula to test this


[ghstack-poisoned]

ezyang commented on aten/src/ATen/NestedTensorImpl.h:
```diff
@@ -169,6 +169,8 @@ struct TORCH_API NestedTensorImpl : public c10::TensorImpl {
   //       we can compute it from `nested_sizes_`
   mutable c10::optional<std::vector<int64_t>> opt_sizes_;
 
+  mutable c10::optional<std::vector<c10::SymInt>> sym_sizes_;
```
This is a pre-existing condition but what's going on with these being mutable fields?

ezyang commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -6371,6 +6371,7 @@
 - func: zeros(SymInt[] size, *, ScalarType? dtype=None, Layout? layout=None, Device? device=None, bool? pin_memory=None) -> Tensor
   dispatch:
     CompositeExplicitAutograd: zeros_symint
+    NestedTensorCPU, NestedTensorCUDA: zeros_nested_symint
```
This is kind of smelly, idk

soulitzer replied:
Oh hmm is the idea that this is bad because its not immediately obvious that only strided NestedTensors go through this path?

soulitzer replied:
We're only going to add as many of these as it takes to get rid of strided NestedTensors specific backward formulas though.

soulitzer committed:
Update on "Factory function and basic .sizes() support for C++ NestedTensor"


This PR:
- updates backend select codegen to also check for keys in singleton ints
- updates C++ NestedTensor to return sizes using singleton ints
- Adds support for C++ NestedTensor factory function to create NestedTensor using metadata on singleton int. Only supports a single ragged dimension.
- Unifies codepath for a particular backward formula to test this


[ghstack-poisoned]

ezyang commented on aten/src/ATen/NestedTensorImpl.cpp:
```diff
@@ -277,7 +278,27 @@ IntArrayRef NestedTensorImpl::sizes_custom() const {
+          c10::SymNode(c10::make_intrusive<c10::NestedIntSymNodeImpl>(-1, -1, std::move(vec), -1, c10::NestedTensorVariant::CPP)));
+      }
+    }
+    sym_sizes_ = std::move(sym_sizes);
```
This doesn't look thread safe

ezyang replied:
Why isn't sym_sizes_ just appropriately populated at construction time

soulitzer replied:
Hmm, we can probably do this.

ezyang commented on tools/autograd/derivatives.yaml:
```diff
@@ -1553,12 +1553,8 @@
-    AutogradNestedTensor:
-      self: _nested_split_with_sizes_backward(grads, split_sizes, dim, at::native::get_nested_tensor_impl(self)->get_nested_sizes(), self.options())
+  self: split_with_sizes_backward(grads, split_sizes, dim, self.sym_sizes(), self.options())
+  result: auto_linear
```
nice

ezyang commented on aten/src/ATen/native/nested/NestedTensorUtils.h:
```diff
@@ -315,7 +316,7 @@ inline Tensor wrap_tensor_node(
           pin_memory);
   if (tensor_node.degree() == 0) {
-    return wrap_buffer(ones({0}, dtype, layout, device), ones({}));
+    return wrap_buffer(ones({0}, at::kLong, at::kStrided, device), ones({}));
```
This seems like you're just ignoring dtype and layout entirely. What's going on?

soulitzer replied:
was trying to fix the fix the size returning float, but looks like I updated the wrong arg lol

ezyang commented on test/test_nestedtensor.py:
```diff
@@ -385,11 +385,32 @@ def test_numel(self):
+
+        nt = torch.nested.nested_tensor([])
+        reconstructed = torch._C._nested_sizes_from_sym_sizes(nt.size())
+        # TODO: figure out why we are returning floating point sizes
```
seems pretty suspicious lol

soulitzer replied:
fixed

ezyang commented on torchgen/gen.py:
```diff
@@ -946,37 +946,44 @@ def __call__(self, f: NativeFunction) -> Optional[str]:
         if self.target is Target.DEFINITION:
+            for a in native_sig_args:
+                if a.argument.type.is_list_like() and a.argument.type.is_symint_like():  # type: ignore[union-attr]
+                    assert a.name == "size"
```
🤔

ezyang commented on aten/src/ATen/native/nested/NestedTensorUtils.cpp:
```diff
@@ -167,5 +170,36 @@ std::vector<Tensor> split_with_sizes_nested(
+Tensor get_nested_sizes_from_sym_sizes(const c10::SymIntArrayRef& size) {
+  if (size.size() == 1) {
+    auto out = at::ones({}, TensorOptions().dtype(at::kLong));
+    return out;
```
Are you sure you want a scalar here? TBH I'm not really sure what this function does

soulitzer replied:
The sym_size function goes from column 1 to column 2, and this function goes from column 2 back to column 1. (You were right - previous iteration had missed a case.)

```
                              nested size           shape
NT with 0 constituents        torch.tensor(1)       Size([0])
NT with scalar constituents   torch.empty([N, 0])   Size([N])
NT with K-d constituents      N x K tensor          Size([N, j0, j1, ..., jk])
```

soulitzer committed:
Update on "Factory function and basic .sizes() support for C++ NestedTensor"


This PR:
- updates backend select codegen to also check for keys in singleton ints
- updates C++ NestedTensor to return sizes using singleton ints
- Adds support for C++ NestedTensor factory function to create NestedTensor using metadata on singleton int. Only supports a single ragged dimension.
- Unifies codepath for a particular backward formula to test this


[ghstack-poisoned]

soulitzer committed:
Update on "Factory function and basic .sizes() support for C++ NestedTensor"


This PR:
- updates backend select codegen to also check for keys in singleton ints
- updates C++ NestedTensor to return sizes using singleton ints
- Adds support for C++ NestedTensor factory function to create NestedTensor using metadata on singleton int. Only supports a single ragged dimension.
- Unifies codepath for a particular backward formula to test this


[ghstack-poisoned]

soulitzer committed:
Update on "Factory function and basic .sizes() support for C++ NestedTensor"


This PR:
- updates backend select codegen to also check for keys in singleton ints
- updates C++ NestedTensor to return sizes using singleton ints
- Adds support for C++ NestedTensor factory function to create NestedTensor using metadata on singleton int. Only supports a single ragged dimension.
- Unifies codepath for a particular backward formula to test this


[ghstack-poisoned]

soulitzer committed:
Update on "Factory function and basic .sizes() support for C++ NestedTensor"


This PR:
- updates backend select codegen to also check for keys in singleton ints
- updates C++ NestedTensor to return sizes using singleton ints
- Adds support for C++ NestedTensor factory function to create NestedTensor using metadata on singleton int. Only supports a single ragged dimension.
- Unifies codepath for a particular backward formula to test this


[ghstack-poisoned]

soulitzer committed:
Update on "Factory function and basic .sizes() support for C++ NestedTensor"


This PR:
- updates backend select codegen to also check for keys in singleton ints
- updates C++ NestedTensor to return sizes using singleton ints
- Adds support for C++ NestedTensor factory function to create NestedTensor using metadata on singleton int. Only supports a single ragged dimension.
- Unifies codepath for a particular backward formula to test this


[ghstack-poisoned]

soulitzer committed:
Update on "Factory function and basic .sizes() support for C++ NestedTensor"


This PR:
- updates backend select codegen to also check for keys in singleton ints
- updates C++ NestedTensor to return sizes using singleton ints
- Adds support for C++ NestedTensor factory function to create NestedTensor using metadata on singleton int. Only supports a single ragged dimension.
- Unifies codepath for a particular backward formula to test this


[ghstack-poisoned]

soulitzer committed:
Update on "Factory function and basic .sizes() support for C++ NestedTensor"


This PR:
- updates backend select codegen to also check for keys in singleton ints
- updates C++ NestedTensor to return sizes using singleton ints
- Adds support for C++ NestedTensor factory function to create NestedTensor using metadata on singleton int. Only supports a single ragged dimension.
- Unifies codepath for a particular backward formula to test this


[ghstack-poisoned]

soulitzer committed:
Update on "Factory function and basic .sizes() support for C++ NestedTensor"


This PR:
- updates backend select codegen to also check for keys in singleton ints
- updates C++ NestedTensor to return sizes using singleton ints
- Adds support for C++ NestedTensor factory function to create NestedTensor using metadata on singleton int. Only supports a single ragged dimension.
- Unifies codepath for a particular backward formula to test this


[ghstack-poisoned]

soulitzer committed:
Update on "Factory function and basic .sizes() support for C++ NestedTensor"


This PR:
- updates backend select codegen to also check for keys in singleton ints
- updates C++ NestedTensor to return sizes using singleton ints
- Adds support for C++ NestedTensor factory function to create NestedTensor using metadata on singleton int. Only supports a single ragged dimension.
- Unifies codepath for a particular backward formula to test this


[ghstack-poisoned]

soulitzer committed:
Update on "Factory function and basic .sizes() support for C++ NestedTensor"


This PR:
- updates backend select codegen to also check for keys in singleton ints
- updates C++ NestedTensor to return sizes using singleton ints
- Adds support for C++ NestedTensor factory function to create NestedTensor using metadata on singleton int. Only supports a single ragged dimension.
- Unifies codepath for a particular backward formula to test this


[ghstack-poisoned]

soulitzer committed:
Update on "Factory function and basic .sizes() support for C++ NestedTensor"


This PR:
- updates backend select codegen to also check for keys in singleton ints
- updates C++ NestedTensor to return sizes using singleton ints
- Adds support for C++ NestedTensor factory function to create NestedTensor using metadata on singleton int. Only supports a single ragged dimension.
- Unifies codepath for a particular backward formula to test this


[ghstack-poisoned]
